src/org/freedesktop/gstreamer/Pad.java

```diff
@@ -318,9 +328,45 @@ public void block(final Runnable callback) {
             public PadProbeReturn eventReceived(Pad pad, Event event) {
                 callback.run();
                 pad.removeCallback(EVENT_PROBE.class, this);
-                return PadProbeReturn.DROP;
+                return PadProbeReturn.REMOVE;
             }
-        }, GstPadAPI.GST_PAD_PROBE_TYPE_BLOCKING | GstPadAPI.GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM);
+        }, GstPadAPI.GST_PAD_PROBE_TYPE_IDLE);
+    }
+
+    /**
+     * Run a supplier (a callback which returns a value) under a blocked state and wait for the runnable to complete. Return the value to
+     * the waiting thread.
+     * The pad will remain blocked for as long as the callback is active.
+     * The callback may be called from this thread or another thread.
+     *
+     * @param supplier The code to run when pad is blocked
+     */
+    public <T> T blockAndWait(final Supplier<T> supplier) {
+        CompletableFuture<T> future = new CompletableFuture<>();
+        block(() -> {
+            T result = supplier.get();
+            future.complete(result);
+        });
+
+        try {
+            return future.get(1, TimeUnit.MINUTES);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Run a runnable under a blocked state and wait for the runnable to complete.
+     * The pad will remain blocked for as long as the callback is active.
+     * The callback may be called from this thread or another thread.
+     *
+     * @param callback The code to run when pad is blocked
+     */
+    public void blockAndWait(final Runnable callback) {
+        blockAndWait(() -> {
+            callback.run();
+            return null;
+        });
```

Comment:
I'd rather not add these. They are things better implemented outside of the core library, and not sure about the hardcoded assumptions - eg. timeout. Probably better achieved by passing a `FutureTask` into `block()` / `onIdle()` and calling `get()` explicitly.

Reply:
I think that this is a fairly common use-case, for example when changing the Pipeline from a GUI Thread or from a Network Thread. In C there is no well defined way to synchronize these Paths, but in Java there is.

The hardcoded Timeout should probably better be a Parameter with a Default-Overload.

Reply:
It's a fairly common way to cause problems too! :smile: And encouraging people to block their GUI or network thread is not high on my agenda.  Considering people can easily do, if they really need this and know there's no deadlock -

```
var task = new FutureTask(() -> connectSomething() ), null);
pad.block(task);
task.get(40, TimeUnit.DAYS);
```

Reply:
Ok, I can see that this API might lead to unexpected Uses. I'm fine with removing it.
I just found that I wrote the same Boilerplate in multiple places around `.block` and chose to add it; but I can see why you do not want this to be part of the API Bindings. I'll remove it from the splitted PRs.